src/Psalm/Internal/Analyzer/ClassLikeAnalyzer.php

```diff
@@ -512,6 +517,57 @@ public static function getTypeFromValue($value): Union
         }
     }
 
+    /**
+     * Gets the Psalm literal type from a particular value
+     *
+     * @param array|scalar|null $value
+     * @throws InvalidArgumentException
+     *
+     */
+    public static function getLiteralTypeFromValue($value, bool $sealed_array = true): Type\Union
```

Comment:
Maybe getLiteralTypeFromScalarValue could be reused or refactored or something?

Reply:
I overlooked that one, I'll take a look at it in a bit. One thing I'm not too happy about is that allowing `array` here could allow a non-scalar to be passed, but I don't think we have a way to type "possibly deeply nested array with only scalars allowed".

Reply:
Done. I'm half tempted to try adding something like `TNestedArray` where `nested-array<scalar>` would be like `array<scalar|array<scalar|array...>>`, but I'm not sure how useful it would be in practice.

Reply:
It would be a hell to add I think. It means you'll have to implement the logic to compare any array to the nested array.
That means:
- can any array be equal to the nested one
- can any array be contained in the nested and the other way around
- nested assertions probably

etc...